### DISCUSSION TITLE
How to parallelly process all logits of a batch of graphs

### DISCUSSION THREAD

**Discussion by vietngth** [pyg-team/pytorch_geometric#8548] [Q&A] [answered]
I want to process a batch of graphs, where the graphs are stacked in a big matrix following the documentation: https://pytorch-geometric.readthedocs.io/en/latest/advanced/batching.html

The differences are:

1.  My label sets have the sizes of adjacency matrices. This means each entry (i, j) is predicted by concatenating/dot products of each pair of node embeddings of node (i, j). Note that the label matrices are not symmetric.
2. For each node attributes set of a graph, I need to materialize a matrix of the adjacency matrix size, where each entry is a pair-wise concatenating/dot products of node embeddings in order to compare with the labels in 1)

To illustrate the idea better, here is a figure:

![label_predictions_page-0001](https://github.com/pyg-team/pytorch_geometric/assets/44941697/31bb9541-ee5e-4f34-87a1-2669366240f7)

The final goal is to feed all logits at-once in the block-diagonal matrix form, and compare them with the labels using binary cross entropy loss. Could you help me with the predict() function? 

```python
class GraphModel(pl.LightningModule):
    def __init__(self, **kargs): 
        ...
        # Some GNN layers like GAT
        self.encoder = ...
        
        self.predictor= torch.nn.Sequential(
            Linear(latent_size * 2, decoder_size),
            ReLU(),
            BatchNorm1d(decoder_size),
            Linear(decoder_size, 1),
            ReLU())

    def encode(self, x, edge_attr, edge_index):
        return self.encoder(x, edge_index, edge_attr)

    def predict(self, z, batch_index):
         # TODO: compute pair-wise node embeddings and stack the matrices
         # z: latent vector
```

**Comment by rusty1s** [comment, marked as answer]
I personally think the easiest way to do this would be via stacking of labels in a separate batch dimension and utilize padding (e.g., via `to_dense_batch`). I think you need to do this anyway inside the model, and so there doesn't exist a good reason to not do it already during batch creation. Then, you can directly compare `z` and `y` and compute BCE loss on top, e.g., via:
```def predict(self, z, batch_index, y):
    z, mask = to_dense_batch(z, batch_index)
    z_t = z.transpose(1, 2)
    # Compute a prediction for every edge using `z` and `z_t`:
    pred = ...
    return bce_loss(pred[mask], y[mask])

**Reply by vietngth** [reply]
Hi, thank you for the suggestion. 

From the documentation, the outputs of `z` and `z_t` should be:
```
z shape: (batch_size, max_node, feature_dim)
z_t shape: (batch_size, feature_dim, max_nodes)
```

Here is a pseudo-code after playing around with torch:

```python
z, mask = to_dense_batch(z, batch_index)
z_t = z.transpose(1, 2)
# Compute a prediction for every edge using `z` and `z_t`:
pair_wise_embeddings = z @ z_t # shape (batch_size, max_nodes, max_nodes)
max_nodes = z.size(1)
feature_dim = z.size(2)
# Create block-diagonal matrix by stacking matrices along new dimension
blocks = torch.stack([pair_wise_embeddings]*max_nodes, dim=2)
flatten= blocks.view(-1, max_nodes * feature_dim)
pred = self.predictor(flatten)

# Assume labels are stored in block-diagonal matrix
labels = batch.y      

return bce_loss(pred[mask], y[mask])
``` 

My questions are:
1. Could you clarify the correctness of my implementation?
2. If I store a matrix label for each graph, will they be stacked in a block-diagonal fashion similar to the adjacency matrices? For example, what will be the shape of `y = batch.y`?

**Reply by rusty1s** [reply]
1. I am not sure I understand what `blocks` is trying to do. You should already have the pairwise predictions for every graph in your batch, so this looks like a waste of memory.
2. I would probably store them as `[1, max_nodes, max_nodes]` matrix in `data.y`.

**Reply by vietngth** [reply]
1. My idea is to create `blocks` that stack the pair-wise embeddings of all the graphs to the form of block-diagonal matrix for `X` as mentioned in the above figure.  But maybe I just misunderstand the purpose of stacking in this formula. Performing a dot product on `z`, i.e. $z^T z$ would already yield a tensor shape `[batch_size, max_nodes, max_nodes]`, and I can just compare them directly with the labels without doing the blocks operation. Do I understand this correctly?

2. This point is clear to me, thanks!

**Reply by rusty1s** [reply]
1. Yes :)

**Reply by vietngth** [reply]
Hi, I have some new issues with the linear layers after doing this way.

```python
class Predictor(pl.LightningModule):
    def __init__(self, **kargs):
        super().__init__()
        # init and GNN layers
        ...

        # but the max_node varies from batch to batch
        self.predictor = Linear(batch_size*max_node*max_node, 1)

    def encode(self, x, edge_index, edge_attr):
        return self.encoder(x, edge_index, edge_attr)

    def forward(self, x, edge_index, batch_index, edge_attr):
        z = self.encode(x, edge_index, edge_attr)
        z, mask = to_dense_batch(z, batch_index)
        z_t = z.transpose(1, 2)
        # pair-wise embeddings of size [batch, max_node, max_node]
        pairwise_emb = z @ z_t
        out = self.predictor(pairwise_emb.flatten())
        return out, mask
    ...
    def training_step(self, batch, batch_idx):
        logits, mask = self(batch.x, batch.edge_index, batch.batch, batch.edge_attr)
        y = batch.label
        batch_size = len(batch)
        logits = logits.view(batch_size, y.shape[1], y.shape[1])[mask]
        y = y.view(logits.shape)[mask]
        loss = self.bce(logits, y)
        return loss

```
1. `max_node` is variable from batch to batch, so `in_features` of `Linear` layer cannot be fixed.
2.  Even if `Linear` can handle variable `in_features` (theoretically), the number of parameters for the linear layer would be significantly larger than the parameters in the GNN layers, thus resulting in too complex overall model.

Is there a way to tackle this problem?

**Reply by rusty1s** [reply]
I don't think you need the `predictor` at all. The inner product already gives you a pairwise similarity.

**Reply by vietngth** [reply]
Is it a common practice to use fully-connected layers after learning the node embeddings for classification task? Normally, I would feed the each pair-wise embeddings into some MLPs to get the final predictions. Does using these embeddings directly for the prediction task hurt the performance?

**Reply by rusty1s** [reply]
This depends on your learning task. In your case, I assume your output should be permutation-invariant. An MLP on top of pair-wise predictions will not respect this permutation-invariance.

**Reply by vietngth** [reply]
Thanks for pointing out the permutation invariance point. My current (extremely slow) implementation computes pair-wise node embeddings and calculates the loss for each graph. Instead of using the whole logit matrix, only its triangular upper part is fed into an MLP. Then, the final predictions are mirrored in the inference step to form a complete matrix. I noticed the model performance was better with MLP than using the pair-wise embeddings directly with a sigmoid function.

Does this help with the permutation-invariant point? Since each pair of nodes (i, j) and (j, i) are considered to be the same by this approach.

**Reply by rusty1s** [reply]
Not totally sure how your implementation looks like but if your `Linear` layer depends on the number of nodes in its input, then it will not help with permutation invariance. It might work well for your data since you overfit on certain permutations, but it will likely fail if you randomly shuffle your data. Maybe I am misunderstanding though and permutation invariance is not a requirement. In this case, everything is good :)

**Reply by vietngth** [reply]
thank you for your support, I think it's clear to me now. I'm gonna close the thread here, happy holiday!
